Checking sizes of input arrays

abhinaukumar opened the pull request:
I have added code to check the shapes of the b1, gr, and tp variables in bloch/bloch.py. The shapes of the other two arrays seem to be independent of these variables, and each other, so I haven't checked their shapes.

jtamir commented:
Thanks! Did you run the tests?

jtamir commented on bloch/bloch.py:
```diff
@@ -43,7 +43,20 @@ def bloch(b1, gr, tp, t1, t2, df, dp, mode, mx=None, my=None, mz=None):
+    tp_flag |= (tp.ndim == 1)
+    if tp.ndim == 2:
+        tp_flag |= (tp.shape[0] == 1) and ((tp.shape[1] == 1) or (tp.shape[1] == ntime))
+    else:
+        tp_flag |= False
```
Is this necessary?

abhinaukumar replied:
Technically, no. I just added it because I thought it might be easier to read. I'll remove it.

abhinaukumar replied:
This is no longer a concern after 0aa5b7b.

abhinaukumar commented:
I forgot, sorry. I ran the tests now and they're failing, which actually brings me to a question that I wanted to ask you. When you get a scalar input for `b1` (`NUMBER` type), you "promote" it to a 1x1 array and interpret it as a simulation for one time-step. I think there is ambiguity when we try to promote the gradient array. Suppose we get an array of size (3,). Do we want to interpret it as a (1,3) array or a (3,1) array? Beyond 3, it's clear that we should resize (N,) into (1, N) because the first dimension can be of size at most 3, but for N <= 3, it seemed a little ambiguous to me. So, I was forcing the gradient array to explicitly be 2-dimensional to avoid having to promote it, which is why the tests fail.

jtamir commented:
We don't want to break backwards compatibility. Can you change it to interpret the old way for array of size `(N,)` where `N <= 3` and make a warning about ambiguity and advising to explicitly make it 2d?

abhinaukumar commented:
I should've read the code better. The routines in bloch_processing.py were already dedicated to preprocessing inputs. So, I moved the input checks into those functions and wrote an extra routine for RF. I also added a warning about small 1-dim gradient arrays, but I haven't changed the way it was being treated, so it will be backwards compatible. All the unit tests passed, but there is now one warning because one of the inputs is a 1-dim gradient array of size (3,). This is the desired behaviour, so nothing to worry about.

jtamir commented:
Looks great! Thanks for putting this together.

abhinaukumar commented:
No problem! I'm happy to help in any small way.
